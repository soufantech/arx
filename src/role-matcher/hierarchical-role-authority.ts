export type RoleHierarchyRepresentation = Record<string, string[]>;
import { RoleAuthority } from './role-matcher';
import { ArxError } from '../errors';

export class HierarchyMappingError extends ArxError {}

type HierarchyMapping = Record<string, string[]>;

export class HierarchicalRoleAuthority implements RoleAuthority {
  private mapping: HierarchyMapping;

  constructor(representation: RoleHierarchyRepresentation) {
    this.mapping = this.map(representation);
  }

  public setHierarchy(representation: RoleHierarchyRepresentation): void {
    this.mapping = this.map(representation);
  }

  public getReachableRoles(grantedRoles: string[]): string[] {
    return Array.from(
      grantedRoles.reduce((set, grantedRole) => {
        // XXX: maybe constructing new set would be more performant?
        (this.mapping[grantedRole] ?? []).forEach((r) => set.add(r));

        return set;
      }, new Set<string>([])),
    );
  }

  public getMapping(): HierarchyMapping {
    return this.mapping;
  }

  private map(representation: RoleHierarchyRepresentation): HierarchyMapping {
    const mapping: HierarchyMapping = {};

    function traverse(role: string, path: string[]): string[] {
      const subjects = representation[role];

      if (mapping[role] == null) {
        mapping[role] = [role];
      }

      if (subjects == null) {
        return [];
      }

      if (path.includes(role)) {
        path.push(role);
        const pathGraphic = path
          .map((r) => {
            return r === role ? `[${r}]` : r;
          })
          .join(' > ');
        throw new HierarchyMappingError(
          `Circular reference error: role [${role}] is backreferenced in path ${pathGraphic}`,
        );
      }

      path.push(role);

      mapping[role] = [role, ...subjects];

      for (const subject of subjects) {
        for (const subsub of traverse(subject, [...path])) {
          if (!mapping[role].includes(subsub)) {
            mapping[role].push(subsub);
          }
        }
      }

      return mapping[role];
    }

    for (const node of Object.keys(representation)) {
      traverse(node, []);
    }

    return mapping;
  }
}
