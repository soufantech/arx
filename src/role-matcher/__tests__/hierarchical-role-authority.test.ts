import { HierarchicalRoleAuthority } from '../hierarchical-role-authority';

describe('HierarchicalRoleAuthority', () => {
  describe('mapping', () => {
    it('detects a circular reference when a role is backreferenced directly in its own child nodes.', () => {
      const CHILD_NODES_CIRC_REF_1 = {
        '1': ['1', '2'],
      };

      const CHILD_NODES_CIRC_REF_2 = {
        '1': ['2', '1'],
        '2': ['3'],
      };

      expect(() => {
        new HierarchicalRoleAuthority(CHILD_NODES_CIRC_REF_1);
      }).toThrow(
        'Circular reference error: role [1] is backreferenced in path [1] > [1]',
      );

      expect(() => {
        new HierarchicalRoleAuthority(CHILD_NODES_CIRC_REF_2);
      }).toThrow(
        'Circular reference error: role [1] is backreferenced in path [1] > [1]',
      );
    });

    it('detects a circular reference when roles are mutually backreferenced.', () => {
      const MUTUAL_CIRC_REF_1 = {
        '1': ['2'],
        '2': ['1'],
      };

      const MUTUAL_CIRC_REF_2 = {
        '1': ['3', '2'],
        '2': ['4', '1'],
      };

      expect(() => {
        new HierarchicalRoleAuthority(MUTUAL_CIRC_REF_1);
      }).toThrow(
        'Circular reference error: role [1] is backreferenced in path [1] > 2 > [1]',
      );

      expect(() => {
        new HierarchicalRoleAuthority(MUTUAL_CIRC_REF_2);
      }).toThrow(
        'Circular reference error: role [1] is backreferenced in path [1] > 2 > [1]',
      );
    });

    it('detects a circular reference when a role is backreferenced deep in the graph.', () => {
      const DEEP_CIRC_REF = {
        '1': ['2', '3'],
        '2': ['3', '4'],
        '4': ['5'],
        '5': ['6', '7', '8'],
        '8': ['9', '2', '10'], // '2' is backreferenced.
      };

      expect(() => {
        new HierarchicalRoleAuthority(DEEP_CIRC_REF);
      }).toThrow(
        'Circular reference error: role [2] is backreferenced in path 1 > [2] > 4 > 5 > 8 > [2]',
      );
    });

    it('detects a circular reference when a role is backreferenced indirectly.', () => {
      const INDIRECT_CIRC_REF_1 = {
        '1': ['2'],
        '2': ['3'],
        '3': ['1'],
      };

      const INDIRECT_CIRC_REF_2 = {
        '1': ['2'],
        '2': ['3'],
        '3': ['4', '1'],
      };

      expect(() => {
        new HierarchicalRoleAuthority(INDIRECT_CIRC_REF_1);
      }).toThrow(
        'Circular reference error: role [1] is backreferenced in path [1] > 2 > 3 > [1]',
      );

      expect(() => {
        new HierarchicalRoleAuthority(INDIRECT_CIRC_REF_2);
      }).toThrow(
        'Circular reference error: role [1] is backreferenced in path [1] > 2 > 3 > [1]',
      );
    });

    it('maps a line-shaped directed acyclic graph.', () => {
      const LINE_DAG = {
        '1': ['2'],
        '2': ['3'],
        '3': ['4'],
        '4': ['5'],
      };

      const authority = new HierarchicalRoleAuthority(LINE_DAG);

      expect(authority.getMapping()).toMatchSnapshot();
    });

    it('maps a tree-shaped directed acyclic graph.', () => {
      const TREE_DAG = {
        '1': ['2', '3'],
        '2': ['4', '5'],
        '3': ['6', '7'],
        '7': ['8'],
      };

      expect(
        new HierarchicalRoleAuthority(TREE_DAG).getMapping(),
      ).toMatchSnapshot();
    });

    it('maps a diamond-shaped directed acyclic graph.', () => {
      const DIAMOND_DAG = {
        '1': ['2', '3'],
        '2': ['4'],
        '3': ['4'], // Both '2' and '3' reference '4'
      };

      expect(
        new HierarchicalRoleAuthority(DIAMOND_DAG).getMapping(),
      ).toMatchSnapshot();
    });

    it('maps a baloon-shaped directed acyclic graph.', () => {
      // a baloon is like a diamond, but with a line in its end.
      const BALOON_DAG = {
        '1': ['2', '3'],
        '2': ['4'],
        '3': ['4'],
        '4': ['5'], // <- a diamond doesn't have this line.
      };

      expect(
        new HierarchicalRoleAuthority(BALOON_DAG).getMapping(),
      ).toMatchSnapshot();
    });

    it('maps a stick-figure-shaped directed acyclic graph.', () => {
      const STICK_FIGURE_DAG = {
        '1': ['2', '3'],
        '2': ['4'],
        '3': ['4'],
        '4': ['5', '6', '7'],
        '7': ['8'],
        '8': ['9', '10'],
      };

      expect(
        new HierarchicalRoleAuthority(STICK_FIGURE_DAG).getMapping(),
      ).toMatchSnapshot();
    });

    it('maps DAGs containing redundant roles.', () => {
      const REDUNDANT_TREE_DAG = {
        '1': ['2', '3', '4', '5', '6', '7'],
        '2': ['4', '5', '8'],
        '3': ['6', '7'],
      };

      expect(
        new HierarchicalRoleAuthority(REDUNDANT_TREE_DAG).getMapping(),
      ).toMatchSnapshot();
    });

    it('maps DAGs containing empty role arrays.', () => {
      const EMPTY_RANKLINE_DAG_1 = {
        '1': [],
      };

      const EMPTY_RANKLINE_DAG_2 = {
        '1': [],
        '2': ['4'],
        '3': [],
      };

      expect(
        new HierarchicalRoleAuthority(EMPTY_RANKLINE_DAG_1).getMapping(),
      ).toMatchSnapshot();

      expect(
        new HierarchicalRoleAuthority(EMPTY_RANKLINE_DAG_2).getMapping(),
      ).toMatchSnapshot();
    });

    it('maps empty directed acyclic graphs.', () => {
      const EMPTY_DAG = {};

      expect(
        new HierarchicalRoleAuthority(EMPTY_DAG).getMapping(),
      ).toMatchSnapshot();
    });

    it('maps deeply nested directed acyclic graphs.', () => {
      // 3' is pointed by '1', '2' and '8' (even though 8 is referenced indirectly by '1's and '2's chain).
      const DEEPLY_NESTED_DAG = {
        '1': ['2', '3'],
        '2': ['3', '4'],
        '4': ['5'],
        '5': ['6', '7', '8'],
        '8': ['9', '3', '10'],
      };

      expect(
        new HierarchicalRoleAuthority(DEEPLY_NESTED_DAG).getMapping(),
      ).toMatchSnapshot();
    });

    it('maps a DAG containing different DAGs.', () => {
      // in this test we have tree trees:
      const TREE_DAG_1 = {
        '1': ['2', '3'],
        '2': ['4', '5'],
        '3': ['6'],
      };

      const TREE_DAG_2 = {
        A: ['B', 'C'],
        B: ['D', 'E'],
        C: ['F'],
      };

      const TREE_DAG_3 = {
        a: ['b', 'c'],
        b: ['d', 'e'],
        c: ['f'],
      };

      const MULTIPLE_TREE_DAG = {
        ...TREE_DAG_1,
        ...TREE_DAG_2,
        ...TREE_DAG_3,
      };

      expect(
        new HierarchicalRoleAuthority(MULTIPLE_TREE_DAG).getMapping(),
      ).toMatchSnapshot();
    });
  });
});
