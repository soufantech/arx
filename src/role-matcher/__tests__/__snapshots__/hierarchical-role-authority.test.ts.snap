// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HierarchicalRoleAuthority mapping maps DAGs containing empty ranklines. 1`] = `
Object {
  "1": Array [
    "1",
  ],
}
`;

exports[`HierarchicalRoleAuthority mapping maps DAGs containing empty ranklines. 2`] = `
Object {
  "1": Array [
    "1",
  ],
  "2": Array [
    "2",
    "4",
  ],
  "3": Array [
    "3",
  ],
  "4": Array [
    "4",
  ],
}
`;

exports[`HierarchicalRoleAuthority mapping maps DAGs containing redundant ranks. 1`] = `
Object {
  "1": Array [
    "1",
    "2",
    "3",
    "4",
    "5",
    "6",
    "7",
    "8",
  ],
  "2": Array [
    "2",
    "4",
    "5",
    "8",
  ],
  "3": Array [
    "3",
    "6",
    "7",
  ],
  "4": Array [
    "4",
  ],
  "5": Array [
    "5",
  ],
  "6": Array [
    "6",
  ],
  "7": Array [
    "7",
  ],
  "8": Array [
    "8",
  ],
}
`;

exports[`HierarchicalRoleAuthority mapping maps a DAG containing different DAGs. 1`] = `
Object {
  "1": Array [
    "1",
    "2",
    "3",
    "4",
    "5",
    "6",
  ],
  "2": Array [
    "2",
    "4",
    "5",
  ],
  "3": Array [
    "3",
    "6",
  ],
  "4": Array [
    "4",
  ],
  "5": Array [
    "5",
  ],
  "6": Array [
    "6",
  ],
  "A": Array [
    "A",
    "B",
    "C",
    "D",
    "E",
    "F",
  ],
  "B": Array [
    "B",
    "D",
    "E",
  ],
  "C": Array [
    "C",
    "F",
  ],
  "D": Array [
    "D",
  ],
  "E": Array [
    "E",
  ],
  "F": Array [
    "F",
  ],
  "a": Array [
    "a",
    "b",
    "c",
    "d",
    "e",
    "f",
  ],
  "b": Array [
    "b",
    "d",
    "e",
  ],
  "c": Array [
    "c",
    "f",
  ],
  "d": Array [
    "d",
  ],
  "e": Array [
    "e",
  ],
  "f": Array [
    "f",
  ],
}
`;

exports[`HierarchicalRoleAuthority mapping maps a baloon-shaped directed acyclic graph. 1`] = `
Object {
  "1": Array [
    "1",
    "2",
    "3",
    "4",
    "5",
  ],
  "2": Array [
    "2",
    "4",
    "5",
  ],
  "3": Array [
    "3",
    "4",
    "5",
  ],
  "4": Array [
    "4",
    "5",
  ],
  "5": Array [
    "5",
  ],
}
`;

exports[`HierarchicalRoleAuthority mapping maps a diamond-shaped directed acyclic graph. 1`] = `
Object {
  "1": Array [
    "1",
    "2",
    "3",
    "4",
  ],
  "2": Array [
    "2",
    "4",
  ],
  "3": Array [
    "3",
    "4",
  ],
  "4": Array [
    "4",
  ],
}
`;

exports[`HierarchicalRoleAuthority mapping maps a line-shaped directed acyclic graph. 1`] = `
Object {
  "1": Array [
    "1",
    "2",
    "3",
    "4",
    "5",
  ],
  "2": Array [
    "2",
    "3",
    "4",
    "5",
  ],
  "3": Array [
    "3",
    "4",
    "5",
  ],
  "4": Array [
    "4",
    "5",
  ],
  "5": Array [
    "5",
  ],
}
`;

exports[`HierarchicalRoleAuthority mapping maps a stick-figure-shaped directed acyclic graph. 1`] = `
Object {
  "1": Array [
    "1",
    "2",
    "3",
    "4",
    "5",
    "6",
    "7",
    "8",
    "9",
    "10",
  ],
  "10": Array [
    "10",
  ],
  "2": Array [
    "2",
    "4",
    "5",
    "6",
    "7",
    "8",
    "9",
    "10",
  ],
  "3": Array [
    "3",
    "4",
    "5",
    "6",
    "7",
    "8",
    "9",
    "10",
  ],
  "4": Array [
    "4",
    "5",
    "6",
    "7",
    "8",
    "9",
    "10",
  ],
  "5": Array [
    "5",
  ],
  "6": Array [
    "6",
  ],
  "7": Array [
    "7",
    "8",
    "9",
    "10",
  ],
  "8": Array [
    "8",
    "9",
    "10",
  ],
  "9": Array [
    "9",
  ],
}
`;

exports[`HierarchicalRoleAuthority mapping maps a tree-shaped directed acyclic graph. 1`] = `
Object {
  "1": Array [
    "1",
    "2",
    "3",
    "4",
    "5",
    "6",
    "7",
    "8",
  ],
  "2": Array [
    "2",
    "4",
    "5",
  ],
  "3": Array [
    "3",
    "6",
    "7",
    "8",
  ],
  "4": Array [
    "4",
  ],
  "5": Array [
    "5",
  ],
  "6": Array [
    "6",
  ],
  "7": Array [
    "7",
    "8",
  ],
  "8": Array [
    "8",
  ],
}
`;

exports[`HierarchicalRoleAuthority mapping maps deeply nested directed acyclic graphs. 1`] = `
Object {
  "1": Array [
    "1",
    "2",
    "3",
    "4",
    "5",
    "6",
    "7",
    "8",
    "9",
    "10",
  ],
  "10": Array [
    "10",
  ],
  "2": Array [
    "2",
    "3",
    "4",
    "5",
    "6",
    "7",
    "8",
    "9",
    "10",
  ],
  "3": Array [
    "3",
  ],
  "4": Array [
    "4",
    "5",
    "6",
    "7",
    "8",
    "9",
    "3",
    "10",
  ],
  "5": Array [
    "5",
    "6",
    "7",
    "8",
    "9",
    "3",
    "10",
  ],
  "6": Array [
    "6",
  ],
  "7": Array [
    "7",
  ],
  "8": Array [
    "8",
    "9",
    "3",
    "10",
  ],
  "9": Array [
    "9",
  ],
}
`;

exports[`HierarchicalRoleAuthority mapping maps empty directed acyclic graphs. 1`] = `Object {}`;
